x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

```diff
@@ -0,0 +1,508 @@
+// Poller function that is polling every 20s, forever until function is resolved.
+async function poll<T>(
+  fn: () => Promise<T>,
+  retries: number = Infinity,
+  interval: number = 20000
+): Promise<T> {
+  return Promise.resolve()
+    .then(fn)
+    .catch(async function retry(err: any) {
+      if (retries-- > 0)
+        return new Promise((resolve) => setTimeout(resolve, interval)).then(fn).catch(retry);
+      throw err;
+    });
+}
+
+// Method to invoke the create environment API for serverless.
+async function createEnvironment(
+  baseUrl: string,
+  projectName: string,
+  runnerId: string,
+  apiKey: string,
+  onEarlyExit: (msg: string) => void
+): Promise<Environment> {
+  console.log(`${runnerId}: Creating environment ${projectName}...`);
+  let environment = {} as Environment;
+  await axios
+    .post(
+      `${baseUrl}/api/v1/serverless/projects/security`,
+      {
+        name: `${projectName}`,
+        region_id: 'aws-eu-west-1',
+      },
+      {
+        headers: {
+          Authorization: `ApiKey ${apiKey}`,
+          'Content-Type': 'application/json',
+        },
+      }
+    )
+    .then((response) => {
+      environment.name = response.data.name;
+      environment.id = response.data.id;
+      environment.region = response.data.region_id;
+      environment.es_url = `${response.data.endpoints.elasticsearch}:443`;
+      environment.kb_url = `${response.data.endpoints.kibana}:443`;
+      environment.product = response.data.type;
+    })
+    .catch((error) => {
+      onEarlyExit(error);
+    });
+  return environment;
+}
+
+// Method to invoke the delete environment API for serverless.
+async function deleteEnvironment(
+  baseUrl: string,
+  projectId: string,
+  projectName: string,
+  runnerId: string,
+  apiKey: string,
+  onEarlyExit: (msg: string) => void
+) {
+  await axios
+    .delete(`${baseUrl}/api/v1/serverless/projects/security/${projectId}`, {
+      headers: {
+        Authorization: `ApiKey ${apiKey}`,
+      },
+    })
+    .then((response) => {
+      console.log(`${runnerId} : Environment ${projectName} was successfully deleted...`);
+    })
+    .catch((error) => {
+      onEarlyExit(error);
+    });
+}
+
+// Method to reset the credentials for the created environment.
+async function resetCredentials(
+  baseUrl: string,
+  environmentId: string,
+  runnerId: string,
+  apiKey: string
+): Promise<Credentials> {
+  console.log(`${runnerId} : Reseting credentials`);
+  let credentials = {} as Credentials;
+
+  await poll(async () => {
+    return await axios
+      .post(
+        `${baseUrl}/api/v1/serverless/projects/security/${environmentId}/_reset-credentials`,
+        {},
+        {
+          headers: {
+            Authorization: `ApiKey ${apiKey}`,
+          },
+        }
+      )
+      .then((response) => {
+        credentials.password = response.data.password;
+        credentials.username = response.data.username;
+      })
+      .catch((error) => {
+        throw error;
```

Comment:
Have we done a thorough test of all these error conditions?  Logged axios errors will include authorization headers in the `config`, `request`, and `response` properties.  It would be safest to assume the worst and always overwrite the error object.

Reply:
Fixed with [fea8077e3538651cf0e340deec2c9e4a48b23985](https://github.com/elastic/kibana/commit/fea8077e3538651cf0e340deec2c9e4a48b23985)

Great catch. I changed it to reraise an error with text : statusCode + data only.